L1Trigger/TrackerTFP/src/TrackFindingProcessor.cc

```diff
@@ -0,0 +1,264 @@
+    for (int channel = 0; channel < (int)inputs.size(); channel++) {
+      deque<Track*>& input = inputs[channel];
+      deque<PartialFrameTrack> stack;
+      deque<FrameTrack> output;
+      // clock accurate firmware emulation, each while trip describes one clock tick, one stub in and one stub out per tick
+      while (!input.empty() || !stack.empty()) {
+        output.emplace_back(FrameTrack());
+        FrameTrack& frame = output.back();
+        Track* track = pop_front(input);
+        if (track)
+          for (const PartialFrame& pf : track->partials_)
+            stack.emplace_back(track->ttTrackRef_, pf);
+        TTBV ttBV;
+        for (int i = 0; i < partial_out; i++) {
+          if (stack.empty()) {
+            ttBV += TTBV(0, partial_width);
+            continue;
+          }
+          const PartialFrameTrack& pft = stack.front();
+          frame.first = pft.first;
+          ttBV += TTBV(pft.second.to_string());
+          stack.pop_front();
+        }
+        frame.second = ttBV.bs();
+      }
+      // perorm truncation
+      if (enableTruncation_ && (int)output.size() > setup_->numFramesIOHigh())
+        output.resize(setup_->numFramesIOHigh());
+      outputs[channel] = StreamTrack(output.begin(), output.end());
+    }
+  }
+
+  // produce TTTracks
+  void TrackFindingProcessor::produce(const StreamsTrack& inputs, TTTracks& outputs) const {
+    // collect input TTTrackRefs
+    vector<TTTrackRef> ttTrackRefs;
+    ttTrackRefs.reserve(tracks_.size());
+    const TTTrack<Ref_Phase2TrackerDigi_>* last = nullptr;
+    for (const StreamTrack& stream : inputs) {
+      for (const FrameTrack& frame : stream) {
+        const TTTrackRef& ttTrackRef = frame.first;
+        if (frame.first.isNull() || last == ttTrackRef.get())
+          continue;
+        last = ttTrackRef.get();
+        ttTrackRefs.push_back(ttTrackRef);
+      }
+    }
+    // convert input TTTrackRefs into output TTTracks
+    static const DataFormat& dfZT = dataFormats_->format(Variable::zT, Process::gp);
+    outputs.reserve(ttTrackRefs.size());
+    for (const TTTrackRef& ttTrackRef : ttTrackRefs) {
+      auto match = [&ttTrackRef](const Track& track) { return track.ttTrackRef_ == ttTrackRef; };
+      const auto it = find_if(tracks_.begin(), tracks_.end(), match);
+      // TTTrack conversion
+      const int region = ttTrackRef->phiSector();
+      const double aRinv = -.5 * it->inv2R_;
+      const double aphi = deltaPhi(it->phiT_ - it->inv2R_ * setup_->chosenRofPhi() + region * setup_->baseRegion());
+      const double aTanLambda = it->cot_;
+      const double az0 = it->zT_ - it->cot_ * setup_->chosenRofZ();
+      const double ad0 = ttTrackRef->d0();
+      const double aChi2xyfit = it->chi2rphi_;
+      const double aChi2zfit = it->chi2rz_;
+      const double trkMVA1 = (TTTrack_TrackWord::tqMVABins[it->mva_] + TTTrack_TrackWord::tqMVABins[it->mva_ + 1]) / 2.;
```

Comment:
Won't this crash when the mva_ bin number is already the last one? You can use just ```TTTrack_TrackWord::tqMVABins[it->mva_]``` instead of the bin center as the GTT objects using this variable already assume the bin edge.

Reply:
you are right, I did not expect that something commented as // Bin edges for chi2/dof does not include all bin edges. Fixed as proposed.